acslib/ccure/crud.py

```diff
@@ -228,3 +234,119 @@ def delete(self, record_id: int) -> ACSRequestResponse:
+        }
+
+    def search(
+        self,
+        item_type: ClearanceItemTypes,
+        terms: Optional[list] = None,
+        search_filter: Optional[ClearanceItemFilter] = None
+    ) -> list:
+        self.logger.info("Searching for clearance items")
+        search_filter = search_filter or self.default_search_filter
+        request_json = {
+            "TypeFullName": self.type_longifier[item_type],
+            "pageSize": 100,
+            "pageNumber": 1,
+            "DisplayProperties": search_filter.display_properties,
+            "WhereClause": search_filter.filter(terms)
+        }
+        response = self.connection.request(
+            ACSRequestMethod.POST,
+            request_data=ACSRequestData(
+                url=self.connection.config.base_url
+                + self.connection.config.endpoints.GET_ALL_WITH_CRITERIA,
+                request_json=request_json,
+                headers=self.connection.base_headers
+            )
+        )
+        return response.json
+
+    def count(self, item_type: ClearanceItemTypes) -> int:
+        request_json = {
+            "TypeFullName": self.type_longifier[item_type],
+            "pageSize": 0,
+            "CountOnly": True,
+            "WhereClause": ""
+        }
+        response = self.connection.request(
+            ACSRequestMethod.POST,
+            request_data=ACSRequestData(
+                url=self.connection.config.base_url
+                + self.connection.config.endpoints.GET_ALL_WITH_CRITERIA,
+                request_json=request_json,
+                headers=self.connection.base_headers
+            )
+        )
+        return response.json
+
+    def update(
+            self,
+            item_type: ClearanceItemTypes,
+            item_id: str,
+            update_data: dict
+        ) -> ACSRequestResponse:
+        return self.connection.request(
+            ACSRequestMethod.PUT,
+            request_data=ACSRequestData(
+                url=self.config.base_url + self.config.endpoints.EDIT_OBJECT,
+                params={
+                    "type": self.type_longifier[item_type],
+                    "id": item_id
+                },
+                data=self.connection.encode_data({
+                    "PropertyNames": list(update_data.keys()),
+                    "PropertyValues": list(update_data.values())
+                }),
+                headers=self.connection.base_headers | self.connection.HEADER_FOR_FORM_DATA
+            )
+        )
+
+    def create(
+            self,
+            item_type: ClearanceItemTypes,
+            create_data: ClearanceItemCreateData
+        ) -> ACSRequestResponse:
+        """Create a new clearance item object"""
+        create_data_dict = create_data.model_dump()
+        request_data = {
+            "type": "SoftwareHouse.NextGen.Common.SecurityObjects.iStarController",
+            "ID": 5000,  # TODO where is this number coming from
```

Comment:
That's a good question. An internal code that means "Create a new Object"?

Reply:
Looking into this a little more: 5000 is the ObjectID of the iStarController object that all our doors and elevators are children of. We only have one iStarController.
If organizations can have controllers with different IDs or if they can have more than one controller with doors, then we'll have to add controller ID as an argument or variable

Reply:
We should probably go ahead and do this. That way it isn't a magic value in the code.